File: Binary search/1011. Capacity To Ship Packages Within D Days/main.py
# first try
class Solution:
    def shipWithinDays(self, weights, D):
        l, r = max(weights), sum(weights)
        while l < r:
            mid = (r - l)//2 + l
            day = 1
            load = 0
            for w in weights:
                if (load + w) <= mid:
                    load += w
                else:
                    load = w
                    day += 1
            
            if day > D: # exceed limit days, the ship has to carry more weight
                l = mid + 1
            else:
                r = mid
        return l

weights = [1,2,3,4,5,6,7,8,9,10]
s = Solution()
print(s.shipWithinDays(weights, 5))

# similar to fist solution
class Solution:
    def shipWithinDays(self, weights, D):
        l, r = max(weights), sum(weights)
        while l < r:
            mid = (r - l)//2 + l
            day = 1
            load = 0
            for w in (weights):
                if (load + w) <= mid:
                    load += w
                else:
                    load = w
                    day += 1
                    if day > D:
                        l = mid + 1
                        break
            else:
                r = mid
        return l

File: Binary search/1011. Capacity To Ship Packages Within D Days/go/main.go
package main

// time complexity: O(logN)
// space complexity: O(1)
func shipWithinDays(weights []int, days int) int {
	l, r := Max(weights), Sum(weights)
	minW := r

	for l < r {
		mid := l + (r-l)/2
		spent := spentDay(weights, mid, days)
		if spent == -1 {
			l = mid + 1
		} else {
			if minW > mid {
				minW = mid
			}
			r = mid
		}
	}

	return minW
}

func spentDay(weights []int, limitW int, limitDay int) int {
	day := 1
	load := 0
	for _, v := range weights {
		if load+v > limitW {
			day++
			load = v
			if day > limitDay {
				return -1
			}
		} else {
			load += v
		}
	}

	return day
}

func Max(s []int) int {
	m := -1
	for _, v := range s {
		if v > m {
			m = v
		}
	}
	return m
}

func Sum(s []int) int {
	total := 0
	for _, v := range s {
		total += v
	}
	return total
}
